### ISSUE TITLE
permission problem when same imageset is processed by different users (OpenDroneMap)

### ISSUE THREAD

**Issue by fredlllll** [nmoehrle/mvs-texturing#76]
Hi,

im currently working with OpenDroneMap and i ran into an issue (i can work around it but its annoying) stemming from this library

the concrete point of failure is https://github.com/nmoehrle/mvs-texturing/blob/733ff76734c2eff09850b223d16b005203ed2383/libs/tex/generate_texture_views.cpp#L194

this lead to a permission problem.

example:
user "a" runs ODM with pictureset "pic_a". at some point the above mentioned piece of code is invoked, and images are copied to /tmp. these images are now owned by "a", and do not get deleted after completion

now user "b" also runs ODM with pictureset "pic_a", and your library tries to overwrite the images that were left in /tmp. this leads to a thrown util::FileException and ODM crashes.

i would propose using a seperate random directory in tmp each time the program is run (seperate directory so several users could process the same imageset at the same time, if necessary), and also deleting it after the program is done, so the tmp directory doesnt fill up

i would do it myself, but i am not proficient enough in c++.

if the thing with the directory is too much effort, it would be enough (for my purposes) if the images got deleted after the program is done

**Comment by nmoehrle** [comment]
I think your idea to create a temporary folder and removing it afterwards is the best solution. However, the cross-platform function to create a temporary will not be available until c++17 so I will have to put something into MVE. Luckily @andre-schulz has already worked on something (https://github.com/andre-schulz/mvs-texturing/commit/47706e58f9cdc9fa0a580516c873b35373e19faf).
 I hope to get it done by the weekend.

**Comment by fredlllll** [comment]
no rush. thanks for taking it into consideration

**Comment by dakotabenjamin** [comment]
Did you get this into mve? I can't build mvs-texturing in ODM anymore. Error: 
```
[ 56%] Building CXX object libs/tex/CMakeFiles/tex.dir/global_seam_leveling.cpp.                                                                                                                         o
[ 56%] Building CXX object libs/tex/CMakeFiles/tex.dir/histogram.cpp.o
[ 56%] Building CXX object libs/tex/CMakeFiles/tex.dir/local_seam_leveling.cpp.o
[ 57%] Building CXX object libs/tex/CMakeFiles/tex.dir/material_lib.cpp.o
[ 59%] Building CXX object libs/tex/CMakeFiles/tex.dir/generate_texture_views.cp                                                                                                                         p.o
[ 61%] Building CXX object libs/tex/CMakeFiles/tex.dir/obj_model.cpp.o
/home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_                                                                                                                         views.cpp: In function ‘void tex::from_images_and_camera_files(const string&, st                                                                                                                         d::vector<tex::TextureView>*)’:
/home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_                                                                                                                         views.cpp:154:25: error: ‘image_undistort_bundler’ is not a member of ‘mve::imag                                                                                                                         e’
                 image = mve::image::image_undistort_bundler<uint8_t>(image,
                         ^
/home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_                                                                                                                         views.cpp:154:68: error: expected primary-expression before ‘>’ token
                 image = mve::image::image_undistort_bundler<uint8_t>(image,
                                                                    ^
/home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_                                                                                                                         views.cpp:155:30: warning: left operand of comma operator has no effect [-Wunuse                                                                                                                         d-value]
                     cam_info.flen, cam_info.dist[0], cam_info.dist[1]);
                              ^
/home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_                                                                                                                         views.cpp:155:30: warning: right operand of comma operator has no effect [-Wunus                                                                                                                         ed-value]
/home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_                                                                                                                         views.cpp:155:51: warning: right operand of comma operator has no effect [-Wunus                                                                                                                         ed-value]
                     cam_info.flen, cam_info.dist[0], cam_info.dist[1]);
                                                   ^
[ 63%] Building CXX object libs/tex/CMakeFiles/tex.dir/partition_mesh.cpp.o
libs/tex/CMakeFiles/tex.dir/build.make:206: recipe for target 'libs/tex/CMakeFil                                                                                                                         es/tex.dir/generate_texture_views.cpp.o' failed
make[5]: *** [libs/tex/CMakeFiles/tex.dir/generate_texture_views.cpp.o] Error 1
make[5]: *** Waiting for unfinished jobs....

```

**Comment by dakotabenjamin** [comment]
BTW We are building from the regularization branch.

**Comment by nmoehrle** [comment]
My mistake I updated my MVE fork without checking compatibility in all projects. There was an API change that broke the builds, this is fixed in 4fb2414. I'll rebase the regularization branch accordingly.

**Comment by nmoehrle** [comment]
Be aware that the rebasing will require you to force pull the regularization branch.

**Comment by dakotabenjamin** [comment]
We download the zip of the archive in CMake, so it shouldn't be an issue.

**Comment by nmoehrle** [comment]
Should be resolved with the changes proposed in #89.

**Comment by GitStardust** [comment]
> Did you get this into mve? I can't build mvs-texturing in ODM anymore. Error:
> 
> ```shell
> ubuntu@ubuntu-xenial:~/opendronemap/SuperBuild/src/mvstexturing$ cd ../../build/mvstexturing/
> ubuntu@ubuntu-xenial:~/opendronemap/SuperBuild/build/mvstexturing$ cmake -DCMAKE_BUILD_TYPE:STRING=Release -DCMAKE_INSTALL_PREFIX:PATH=/home/ubuntu/opendronemap/SuperBuild/install/mvstexturing ../../src/mvstexturing/
> -- The C compiler identification is GNU 5.4.0
> -- The CXX compiler identification is GNU 5.4.0
> -- Check for working C compiler: /usr/bin/cc
> -- Check for working C compiler: /usr/bin/cc -- works
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Check for working CXX compiler: /usr/bin/c++
> -- Check for working CXX compiler: /usr/bin/c++ -- works
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- Try OpenMP C flag = [-fopenmp]
> -- Performing Test OpenMP_FLAG_DETECTED
> -- Performing Test OpenMP_FLAG_DETECTED - Success
> -- Try OpenMP CXX flag = [-fopenmp]
> -- Performing Test OpenMP_FLAG_DETECTED
> -- Performing Test OpenMP_FLAG_DETECTED - Success
> -- Found OpenMP: -fopenmp
> -- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.8")
> -- Found PNG: /usr/lib/x86_64-linux-gnu/libpng.so (found version "1.2.54")
> -- Found JPEG: /usr/lib/x86_64-linux-gnu/libjpeg.so
> -- Found TIFF: /usr/lib/x86_64-linux-gnu/libtiff.so (found version "4.0.6")
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /home/ubuntu/opendronemap/SuperBuild/build/mvstexturing
> ubuntu@ubuntu-xenial:~/opendronemap/SuperBuild/build/mvstexturing$ make
> Scanning dependencies of target ext_eigen
> [  1%] Creating directories for 'ext_eigen'
> [  3%] Performing download step (download, verify and extract) for 'ext_eigen'
> -- downloading...
>      src='https://bitbucket.org/eigen/eigen/get/3.2.8.tar.bz2'
>      dst='/home/ubuntu/opendronemap/SuperBuild/build/mvstexturing/elibs/ext_eigen/src/3.2.8.tar.bz2'
>      timeout='none'
> -- [download 0% complete]
> -- [download 1% complete]
> -- [download 2% complete]
> -- [download 3% complete]
> -- [download 4% complete]
> -- [download 5% complete]
> -- [download 6% complete]
> -- [download 7% complete]
> -- [download 8% complete]
> -- [download 9% complete]
> -- [download 10% complete]
> -- [download 11% complete]
> -- [download 12% complete]
> -- [download 13% complete]
> -- [download 14% complete]
> -- [download 15% complete]
> -- [download 16% complete]
> -- [download 17% complete]
> -- [download 18% complete]
> -- [download 19% complete]
> -- [download 20% complete]
> -- [download 21% complete]
> -- [download 22% complete]
> -- [download 23% complete]
> -- [download 24% complete]
> -- [download 25% complete]
> -- [download 26% complete]
> -- [download 27% complete]
> -- [download 28% complete]
> -- [download 29% complete]
> -- [download 30% complete]
> -- [download 31% complete]
> -- [download 32% complete]
> -- [download 33% complete]
> -- [download 34% complete]
> -- [download 35% complete]
> -- [download 36% complete]
> -- [download 37% complete]
> -- [download 38% complete]
> -- [download 39% complete]
> -- [download 40% complete]
> -- [download 41% complete]
> -- [download 42% complete]
> -- [download 43% complete]
> -- [download 44% complete]
> -- [download 45% complete]
> -- [download 46% complete]
> -- [download 47% complete]
> -- [download 48% complete]
> -- [download 49% complete]
> -- [download 50% complete]
> -- [download 51% complete]
> -- [download 52% complete]
> -- [download 53% complete]
> -- [download 54% complete]
> -- [download 55% complete]
> -- [download 56% complete]
> -- [download 57% complete]
> -- [download 58% complete]
> -- [download 59% complete]
> -- [download 60% complete]
> -- [download 61% complete]
> -- [download 62% complete]
> -- [download 63% complete]
> -- [download 64% complete]
> -- [download 65% complete]
> -- [download 66% complete]
> -- [download 67% complete]
> -- [download 68% complete]
> -- [download 69% complete]
> -- [download 70% complete]
> -- [download 71% complete]
> -- [download 72% complete]
> -- [download 73% complete]
> -- [download 74% complete]
> -- [download 75% complete]
> -- [download 76% complete]
> -- [download 77% complete]
> -- [download 78% complete]
> -- [download 79% complete]
> -- [download 80% complete]
> -- [download 81% complete]
> -- [download 82% complete]
> -- [download 83% complete]
> -- [download 84% complete]
> -- [download 85% complete]
> -- [download 86% complete]
> -- [download 87% complete]
> -- [download 88% complete]
> -- [download 89% complete]
> -- [download 90% complete]
> -- [download 91% complete]
> -- [download 92% complete]
> -- [download 93% complete]
> -- [download 94% complete]
> -- [download 95% complete]
> -- [download 96% complete]
> -- [download 97% complete]
> -- [download 98% complete]
> -- [download 99% complete]
> -- [download 100% complete]
> -- downloading... done
> -- verifying file...
>      file='/home/ubuntu/opendronemap/SuperBuild/build/mvstexturing/elibs/ext_eigen/src/3.2.8.tar.bz2'
> -- verifying file... done
> -- extracting...
>      src='/home/ubuntu/opendronemap/SuperBuild/build/mvstexturing/elibs/ext_eigen/src/3.2.8.tar.bz2'
>      dst='/home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/elibs/eigen'
> -- extracting... [tar xfz]
> -- extracting... [analysis]
> -- extracting... [rename]
> -- extracting... [clean up]
> -- extracting... done
> [  5%] No patch step for 'ext_eigen'
> [  7%] No update step for 'ext_eigen'
> [  8%] No configure step for 'ext_eigen'
> [ 10%] No build step for 'ext_eigen'
> [ 12%] No install step for 'ext_eigen'
> [ 14%] Completed 'ext_eigen'
> [ 14%] Built target ext_eigen
> Scanning dependencies of target ext_rayint
> [ 15%] Creating directories for 'ext_rayint'
> [ 17%] Performing download step (git clone) for 'ext_rayint'
> Cloning into 'rayint'...
> Switched to a new branch 'cuda'
> Branch cuda set up to track remote branch cuda from origin.
> [ 19%] No patch step for 'ext_rayint'
> [ 21%] No update step for 'ext_rayint'
> [ 22%] No configure step for 'ext_rayint'
> [ 24%] No build step for 'ext_rayint'
> [ 26%] No install step for 'ext_rayint'
> [ 28%] Completed 'ext_rayint'
> [ 28%] Built target ext_rayint
> Scanning dependencies of target ext_mve
> [ 29%] Creating directories for 'ext_mve'
> [ 31%] Performing download step (git clone) for 'ext_mve'
> Cloning into 'mve'...
> Already on 'master'
> Your branch is up-to-date with 'origin/master'.
> [ 33%] No patch step for 'ext_mve'
> [ 35%] No update step for 'ext_mve'
> [ 36%] No configure step for 'ext_mve'
> [ 38%] Performing build step for 'ext_mve'
> [ 40%] No install step for 'ext_mve'
> [ 42%] Completed 'ext_mve'
> [ 42%] Built target ext_mve
> Scanning dependencies of target mrf
> [ 43%] Building CXX object libs/mrf/CMakeFiles/mrf.dir/gco_graph.cpp.o
> [ 45%] Building CXX object libs/mrf/CMakeFiles/mrf.dir/graph.cpp.o
> [ 47%] Building CXX object libs/mrf/CMakeFiles/mrf.dir/icm_graph.cpp.o
> [ 49%] Building CXX object libs/mrf/CMakeFiles/mrf.dir/lbp_graph.cpp.o
> [ 50%] Linking CXX static library libmrf.a
> [ 50%] Built target mrf
> Scanning dependencies of target tex
> [ 52%] Building CXX object libs/tex/CMakeFiles/tex.dir/build_adjacency_graph.cpp.o
> [ 54%] Building CXX object libs/tex/CMakeFiles/tex.dir/build_obj_model.cpp.o
> [ 56%] Building CXX object libs/tex/CMakeFiles/tex.dir/calculate_data_costs.cpp.o
> [ 57%] Building CXX object libs/tex/CMakeFiles/tex.dir/generate_debug_embeddings.cpp.o
> [ 59%] Building CXX object libs/tex/CMakeFiles/tex.dir/generate_texture_atlases.cpp.o
> [ 61%] Building CXX object libs/tex/CMakeFiles/tex.dir/generate_texture_patches.cpp.o
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_patches.cpp: In function ‘bool tex::fill_hole(const std::vector<long unsigned int, std::allocator<long unsigned int> >&, const UniGraph&, mve::TriangleMesh::ConstPtr, const mve::MeshInfo&, std::vector<std::vector<tex::VertexProjectionInfo> >*, std::vector<std::shared_ptr<TexturePatch> >*)’:
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_patches.cpp:278:14: warning: variable ‘test’ set but not used [-Wunused-but-set-variable]
>          bool test = false;
>               ^
> [ 63%] Building CXX object libs/tex/CMakeFiles/tex.dir/generate_texture_views.cpp.o
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_views.cpp: In function ‘void tex::from_images_and_camera_files(const string&, std::vector<tex::TextureView>*)’:
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_views.cpp:154:25: error: ‘image_undistort_bundler’ is not a member of ‘mve::image’
>                  image = mve::image::image_undistort_bundler<uint8_t>(image,
>                          ^
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_views.cpp:154:68: error: expected primary-expression before ‘>’ token
>                  image = mve::image::image_undistort_bundler<uint8_t>(image,
>                                                                     ^
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_views.cpp:155:30: warning: left operand of comma operator has no effect [-Wunused-value]
>                      cam_info.flen, cam_info.dist[0], cam_info.dist[1]);
>                               ^
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_views.cpp:155:30: warning: right operand of comma operator has no effect [-Wunused-value]
> /home/ubuntu/opendronemap/SuperBuild/src/mvstexturing/libs/tex/generate_texture_views.cpp:155:51: warning: right operand of comma operator has no effect [-Wunused-value]
>                      cam_info.flen, cam_info.dist[0], cam_info.dist[1]);
>                                                    ^
> libs/tex/CMakeFiles/tex.dir/build.make:206: recipe for target 'libs/tex/CMakeFiles/tex.dir/generate_texture_views.cpp.o' failed
> make[2]: *** [libs/tex/CMakeFiles/tex.dir/generate_texture_views.cpp.o] Error 1
> CMakeFiles/Makefile2:242: recipe for target 'libs/tex/CMakeFiles/tex.dir/all' failed
> make[1]: *** [libs/tex/CMakeFiles/tex.dir/all] Error 2
> Makefile:127: recipe for target 'all' failed
> make: *** [all] Error 2
> ```
Hello,i have face this problem too,have you solve it? Please help me !
